cpp/oneapi/dal/backend/primitives/search/search.hpp

```diff
@@ -0,0 +1,146 @@
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/backend/primitives/common.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/backend/primitives/distance.hpp"
+#include "oneapi/dal/backend/primitives/selection.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <typename Float>
+std::int64_t propose_train_block(const sycl::queue& q, std::int64_t width);
+
+template <typename Float>
+std::int64_t propose_query_block(const sycl::queue& q, std::int64_t width);
+
+template <typename Float, typename Impl>
+class callback_base {
+public:
+    using float_t = Float;
+    using derived_t = Impl;
+
+    sycl::event operator()(std::int64_t qb_id,
+                           const ndview<std::int32_t, 2>& indices,
+                           const ndview<Float, 2>& distances,
+                           const event_vector& deps = {}) {
+        return static_cast<Impl*>(this)->run(qb_id, indices, distances, deps);
+    }
+};
+
+template <typename Float, typename Distance>
+class search_temp_objects;
+
+template <typename Float, typename Distance>
+class search_engine {
+    using temp_t = search_temp_objects<Float, Distance>;
+    using temp_ptr_t = temp_t* const;
+    using selc_t = kselect_by_rows<Float>;
+    template <typename CallbackImpl>
+    using call_t = callback_base<Float, CallbackImpl>;
+
+    constexpr static inline std::int64_t selection_sub_blocks = 4;
+
+public:
+    search_engine(sycl::queue& queue, const ndview<Float, 2>& train_data);
+
+    search_engine(sycl::queue& queue, const ndview<Float, 2>& train_data, std::int64_t train_block);
+
+    search_engine(sycl::queue& queue,
+                  const ndview<Float, 2>& train_data,
+                  std::int64_t train_block,
+                  const Distance& distance_instance);
+
+    template <typename CallbackImpl>
+    sycl::event operator()(const ndview<Float, 2>& query_data,
+                           CallbackImpl& callback,
+                           std::int64_t k_neighbors = 1,
+                           const event_vector& deps = {}) const {
+        const auto query_block = propose_query_block<Float>(queue_, query_data.get_dimension(1));
+        return this->operator()(query_data.callback, k_neighbors, query_block, deps);
+    }
+
+    template <typename CallbackImpl>
+    sycl::event operator()(const ndview<Float, 2>& query_data,
+                           CallbackImpl& callback,
+                           std::int64_t query_block,
+                           std::int64_t k_neighbors = 1,
+                           const event_vector& deps = {}) const {
+        selc_t selection = create_selection_objects(query_block, k_neighbors);
+        const uniform_blocking query_blocking(query_data.get_dimension(0), query_block);
+        auto tmp_objs = create_temporary_objects(query_blocking, k_neighbors);
```

Comment:
tmp_objs is temp_ptr_t which is temp_t*. It is created in create_temporary_objects by new operator, but i don't see corresponding delete operator here.

Reply:
Fixed and discussed